pkg/aws/eks/pricing_map.go

```diff
@@ -0,0 +1,167 @@
+package eks
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"strconv"
+	"strings"
+	"sync"
+
+	ec2Types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+)
+
+var (
+	ErrInstanceTypeAlreadyExists = errors.New("instance type already exists in the map")
+	ErrCalculatedWeightedCost    = errors.New("could not calculate the weighted cost")
+)
+
+// StructuredPricingMap collects a map of FamilyPricing structs where the key is the region
+type StructuredPricingMap struct {
+	// Regions is a map of region code to FamilyPricing
+	// key is the region
+	// value is a map of instance type to PriceTiers
+	Regions         map[string]*FamilyPricing
+	InstanceDetails map[string]Attributes
+	m               sync.RWMutex
+}
+
+// FamilyPricing is a map of instance type to a list of PriceTiers where the key is the ec2 compute instance type
+type FamilyPricing struct {
+	Family map[string]*ComputePrices // Each Family can have many PriceTiers
+}
+
+// ComputePrices holds the price of a compute instances CPU and RAM. The price is in USD
+type ComputePrices struct {
+	Cpu float64
+	Ram float64
+}
+
+func NewStructuredPricingMap() *StructuredPricingMap {
+	return &StructuredPricingMap{
+		Regions:         make(map[string]*FamilyPricing),
+		InstanceDetails: make(map[string]Attributes),
+		m:               sync.RWMutex{},
+	}
+}
+
+// GeneratePricingMap accepts a list of ondemand prices and a list of spot prices.
+// The method needs to
+// 1. Parse out the ondemand prices and generate a productTerm map for each instance type
+// 2. Parse out spot prices and use the productTerm map to generate a spot price map
+func (spm *StructuredPricingMap) GeneratePricingMap(ondemandPrices []string, spotPrices []ec2Types.SpotPrice) error {
+	for _, product := range ondemandPrices {
+		var productInfo productTerm
+		if err := json.Unmarshal([]byte(product), &productInfo); err != nil {
+			return err
```

Comment:
I think this scenario stuck out the most because if we're unable to unmarshal one price, we're likely unable to unmarshal others. The other scenarios we skip are less problematic(imo).

Reply:
👍